### DISCUSSION TITLE
Numerous exceptions - recommended setup?

### DISCUSSION THREAD

**Discussion by philippkeller** [miguelgrinberg/python-socketio#1180] [Q&A] [answered]
I've run into numerous issues:

- `'Response' object has no attribute 'status_code'` in `gunicorn/http/wsgi.py in should_close at line 219`
- `'Session is disconnected'` in `engineio/server.py in _get_socket at line 637`

My setup:

- gunicorn==20.1.0
- eventlet==0.30.2 (because the newer versions trigger `ImportError: cannot import name 'ALREADY_HANDLED' from 'eventlet.wsgi'`)
- starting with `gunicorn -- --worker-class eventlet -w 1 app:app`

It all seems very buggy with gunicorn (which didn't have a new version since 2 years!), so I'm tempted to switch to a better server.
What is the recommended setup? I'm also willing to take something other than eventlet.

I [looked in the docs](https://python-socketio.readthedocs.io/en/latest/server.html#eventlet-with-gunicorn) and there are some other mentioned servers like Uvicorn and Daphne but I didn't see which setup was recommended.

**Comment by miguelgrinberg** [comment, marked as answer]
I can't really respond for issues on Gunicorn or in the interaction between Gunicorn and eventlet. If you are unable to find a combination of versions that work for you then you can try any of the other options. It seems you are expecting I will be recommending a single option, but unfortunately that is not possible, since there are many different types of application and use cases. All the options that are documented have been personally verified by myself, so you can pick the one that matches your needs more closely. For example, if you like eventlet, but can't find how to make it work with Gunicorn, then use eventlet alone.

**Reply by philippkeller** [reply]
First: I find it amazing how you are supporting SocketIO over the years. Virtually every question (also on stackoverflow) is answered by you - and this one even within minutes on a Sunday 😲 

Then: Thanks, I'll try out some options. Is socketIO supporting multiple workers? I see that with gunicorns that was not the case.

**Reply by miguelgrinberg** [reply]
Socket.IO supports multiple workers, of course. With Gunicorn as well. The problem is that Gunicorn has a very simplistic load balancer that does not work well with Socket.IO. The workaround is to start several single-worker Gunicorns, and then put a better load balancer in front of all of them, such as nginx. This is also in the docs, start from the section titled "Using Message Queues". And yes, the title for this section isn't great, I have to find a better one. :)
